xla/service/gpu/autotuning/gemm_fusion_autotuner_rocm.cc

```diff
@@ -0,0 +1,306 @@
+
+  for (int num_stages : kNumStages) {
+    for (int tile_m : kBlockSizes) {
+      for (int tile_n : kBlockSizes) {
+        for (int tile_k : kBlockSizes) {
+          const int tile_lhs = tile_m * tile_k;
+          const int tile_rhs = tile_k * tile_n;
+          for (int num_warps : kNumWarps) {
+            // Each thread should read at least one input element.
+            if (num_warps * WarpSize() > std::min(tile_lhs, tile_rhs)) {
+              break;
+            }
+            for (int split_k : kSplitK) {
+              // Split-K autotuning may be disabled by a flag.
+              if (!debug_options_.xla_gpu_enable_split_k_autotuning() &&
+                  split_k > 1) {
+                break;
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+  return configs;
+}
+
+std::vector<TritonGemmConfig> GemmFusionAutotunerImpl::GetDefaultTritonConfigs()
+    const {
+  using Config = TritonGemmConfig;
+  std::vector<Config> configs = {
+        Config(32, 32, 256, 1, 1, 4), Config(64, 32, 32, 16, 1, 4),
+        Config(32, 64, 64, 4, 1, 4),  Config(128, 128, 64, 4, 1, 4),
+        Config(16, 16, 256, 1, 1, 4), Config(16, 128, 32, 16, 1, 4),
+  };
+  return configs;
+}
+
+absl::StatusOr<std::vector<BackendConfig>>
+GemmFusionAutotunerImpl::GenerateConfigs(const HloFusionInstruction& fusion) {
+  const HloDotInstruction* dot =
+      Cast<HloDotInstruction>(hlo_query::GetFirstInstructionWithOpcode(
+          *fusion.called_computations().at(0), HloOpcode::kDot));
+  std::vector<BackendConfig> configs;
+
+  if (!debug_options_.xla_gpu_experimental_disable_binary_libraries()) {
+    // Add cuBLAS reference config, if available.
+    if (algorithm_util::IsSupportedByCublasOrCublasLt(
+            dot->precision_config().algorithm(), GetComputeCapability()) &&
+        !dot->sparse_operands() && IsAutotuningEnabled()) {
+      configs.push_back(CuBlasConfig{});
+    }
+  }
+
+  // Add CustomKernelFusion (Cutlass) configs, if available.
+  // Go through all the instructions in the fusion body try to match them to
+  // a custom kernel fusion pattern.
+  if ((IsFusionKind(fusion, kCustomFusionKind) ||
+       IsFusionKind(fusion, kTritonGemmFusionKind)) &&
+      IsAutotuningEnabled() && !config_.IsDeviceless()) {
+    std::vector<BackendConfig> custom_kernel_fusion_configs =
+        GenerateCustomKernelFusionConfigs(
+            fusion, config_.GetExecutor()->GetDeviceDescription());
+    configs.insert(configs.end(), custom_kernel_fusion_configs.begin(),
+                   custom_kernel_fusion_configs.end());
+  }
+
+  // Add triton configs.
+  TF_ASSIGN_OR_RETURN(std::vector<TritonGemmConfig> triton_configs,
+                      GenerateTritonConfigs(*dot));
+  for (TritonGemmConfig& config : triton_configs) {
+    configs.push_back(std::move(config));
+  }
+  return configs;
+}
+
+absl::StatusOr<std::vector<TritonGemmConfig>>
+GemmFusionAutotunerImpl::GenerateTritonConfigs(const HloDotInstruction& dot) {
```

Comment:
This method is too long, please break it down into multiple methods.

Reply:
Just copied existing corresponding Cuda method. If it necessary, I will break it to several smaller methods.

Reply:
Right, the original method is also too long. If you end up modifying the original (if there's just one left after the proposals) or if you fork it, please break it down. It's good to do minor clean ups as we touch old code.